romancal/lib/tests/test_psf.py

```diff
@@ -0,0 +1,157 @@
+
+        return mod
+
+    return _setup
+
+
+def add_synthetic_sources(
+    image_model,
+    psf_model,
+    true_x,
+    true_y,
+    true_amp,
+    oversample,
+    xname="x_0",
+    yname="y_0",
+):
+    fit_models = []
+
+    # ensure truths are arrays:
+    true_x, true_y, true_amp = (
+        np.atleast_1d(truth) for truth in [true_x, true_y, true_amp]
+    )
+
+    for x, y, amp in zip(true_x, true_y, true_amp):
+        psf = psf_model.copy()
+        psf.parameters = [amp, x, y]
+        fit_models.append(psf)
+
+    synth_image = image_model.data
+    synth_err = image_model.err
+    psf_shape = np.array(psf_model.data.shape[1:]) // oversample
+
+    for fit_model in fit_models:
+        x0 = getattr(fit_model, xname).value
+        y0 = getattr(fit_model, yname).value
+        slc_lg, _ = overlap_slices(synth_image.shape, psf_shape, (y0, x0), mode="trim")
+        yy, xx = np.mgrid[slc_lg]
+        model_data = fit_model(xx, yy) * image_model.data.unit
+        model_err = np.sqrt(model_data.value) * model_data.unit
+        synth_image[slc_lg] += model_data
+        synth_err[slc_lg] = np.sqrt(synth_err[slc_lg] ** 2 + model_err**2)
+
+
+@pytest.mark.parametrize(
+    "dx, dy, true_amp",
+    zip(
+        rng.uniform(-1, 1, n_sources),
+        rng.uniform(-1, 1, n_sources),
+        np.geomspace(10, 10_000, n_sources),
+    ),
+)
+def test_psf_fit(setup_inputs, dx, dy, true_amp, seed=42):
+    # input parameters for PSF model:
+    filt = "F087"
+    detector = "SCA01"
+    oversample = 12
+    fov_pixels = 15
+
+    dir_path = tempfile.gettempdir()
+    filename_prefix = f"psf_model_{filt}"
+    file_path = os.path.join(dir_path, filename_prefix)
+
+    # compute gridded PSF model:
+    psf_model, centroids = create_gridded_psf_model(
+        file_path,
+        filt,
+        detector,
+        oversample=oversample,
+        fov_pixels=fov_pixels,
+        overwrite=False,
+        logging_level="ERROR",
+    )
+
+    # generate an ImageModel
+    image_model = setup_inputs(seed=seed)
+
+    # add synthetic sources to the ImageModel:
+    true_x = image_model_shape[0] / 2 + dx
+    true_y = image_model_shape[1] / 2 + dy
+    add_synthetic_sources(
+        image_model, psf_model, true_x, true_y, true_amp, oversample=oversample
+    )
+
+    if fov_pixels % 2 == 0:
+        fit_shape = (fov_pixels + 1, fov_pixels + 1)
+    else:
+        fit_shape = (fov_pixels, fov_pixels)
+
+    # fit the PSF to the ImageModel:
+    results_table, photometry = fit_psf_to_image_model(
+        image_model=image_model,
+        photometry_cls=PSFPhotometry,
+        psf_model=psf_model,
+        x_init=true_x,
+        y_init=true_y,
+        fit_shape=fit_shape,
+    )
+
+    # difference between input and output, normalized by the
+    # uncertainty. Has units of sigma:
+    delta_x = np.abs(true_x - results_table["x_fit"]) / results_table["x_err"]
+    delta_y = np.abs(true_y - results_table["y_fit"]) / results_table["y_err"]
```

Comment:
Can we also assert something about the size of x_err / y_err?  Maybe x_err < 1 / (amp / (sqrt(noise^2 + amp)) * fudge factor?  I just want to be able to catch failure modes like "all of the errors are very large, so we have small delta_x / delta_y but did a poor fit?  I'm trying to use the rule-of-thump x_err ~ FWHM / (S/N), which doesn't apply for this undersampled data but isn't terrible, and using 1 pix for the FWHM, and doing a bad approximation to the S/N.